src/nrniv/netpar.cpp

```diff
@@ -1285,6 +1301,35 @@ void BBS::netpar_solve(double tstop) {
 	tstopunset;
 }
 
+int nrnthread_all_spike_vectors_return(std::vector<double>& spiketvec, std::vector<int>& spikegidvec){
+    assert(spiketvec.size() == spikegidvec.size());
+    if( spiketvec.size() ) {
+        /* Optimisation:
+         *  if all_spiketvec and all_spikegidvec are set (pc.spike_record with gid=-1 was called)
+         *  and they are still reference counted (obj_->refcount), then copy over incoming vectors.
+         */
+        if (all_spiketvec != NULL && all_spiketvec->obj_ != NULL  && all_spiketvec->obj_->refcount > 0 &&
+            all_spikegidvec != NULL && all_spikegidvec->obj_ != NULL && all_spikegidvec->obj_->refcount > 0) {
+
+            all_spiketvec->resize(spiketvec.size());
+            all_spikegidvec->resize(spikegidvec.size());
+            for (int i = 0; i < all_spiketvec->capacity(); ++i) {
+                all_spiketvec->elem(i) = spiketvec[i];
+                all_spikegidvec->elem(i) = spikegidvec[i];
+            }
+        }else{ // different underlying vectors for PreSyns
+            for (int i = 0; i < spikegidvec.size(); ++i ) {
+                PreSyn* ps;
+                if (gid2out_->find(spikegidvec[i], ps)) {
```

Comment:
Thinking about error handling(see comment below), this `find` should not fail.
It might be a good candidate enabling writing `out.dat` (`return 0`)

Reply:
I think you have all the important cases worked out. The only edge case is where the NEURON user asks to record spikes from a PreSyn without an output_index_ and so CoreNEURON will not store those (negative coded) gids.
Although formally possible, I don't think it is worth worrying about at this time.